test/e2e/network/service.go

```diff
@@ -1883,3 +1899,81 @@ func execSourceipTest(f *framework.Framework, c clientset.Interface, ns, nodeNam
+	})
+	Expect(err).NotTo(HaveOccurred())
+	defer func() {
+		framework.StopServeHostnameService(cs, f.InternalClientset, ns, serviceName)
+	}()
+	jig := framework.NewServiceTestJig(cs, serviceName)
+	svc, err := jig.Client.CoreV1().Services(ns).Get(serviceName, metav1.GetOptions{})
+	Expect(err).NotTo(HaveOccurred())
+	var svcIpPort string
+	if serviceType == v1.ServiceTypeNodePort {
+		nodeIP := framework.PickNodeIP(jig.Client)
+		svcIpPort = net.JoinHostPort(nodeIP, strconv.Itoa(int(svc.Spec.Ports[0].NodePort)))
+	} else {
+		svcIpPort = net.JoinHostPort(svc.Spec.ClusterIP, strconv.Itoa(servicePort))
+	}
+
+	execPodName := framework.CreateExecPodOrFail(cs, ns, "execpod-affinity", nil)
+	defer func() {
+		framework.Logf("Cleaning up the exec pod")
+		err := cs.CoreV1().Pods(ns).Delete(execPodName, nil)
+		Expect(err).NotTo(HaveOccurred())
+	}()
+	execPod, err := cs.CoreV1().Pods(ns).Get(execPodName, metav1.GetOptions{})
+	Expect(err).NotTo(HaveOccurred())
+
+	Expect(framework.TestAffinityFromPod(execPod, svcIpPort, false)).To(BeTrue(), "Affinity shouldn't hold or connection timeout.")
+	svc = jig.UpdateServiceOrFail(svc.Namespace, svc.Name, func(svc *v1.Service) {
+		svc.Spec.SessionAffinity = v1.ServiceAffinityClientIP
+	})
+	Expect(framework.TestAffinityFromPod(execPod, svcIpPort, true)).To(BeTrue(), "Affinity didn't hold or connection timeout.")
+}
+
+func execAffinityTestForLBService(f *framework.Framework, cs clientset.Interface, externalTrafficPolicy v1.ServiceExternalTrafficPolicyType) {
+	ns := f.Namespace.Name
+	numPods, servicePort := 3, 80
+	serviceName := "service"
+
+	By("creating service in namespace " + ns)
+	_, _, err := framework.StartCustomizedServeHostnameService(cs, f.InternalClientset, ns, serviceName, servicePort, numPods, func(svc *v1.Service) {
+		svc.Spec.Type = v1.ServiceTypeLoadBalancer
+		svc.Spec.ExternalTrafficPolicy = v1.ServiceExternalTrafficPolicyTypeLocal
+	})
+	Expect(err).NotTo(HaveOccurred())
+	jig := framework.NewServiceTestJig(cs, serviceName)
+	By("waiting for loadbalancer for service " + ns + "/" + serviceName)
+	svc := jig.WaitForLoadBalancerOrFail(ns, serviceName, framework.LoadBalancerCreateTimeoutDefault)
+	jig.SanityCheckService(svc, v1.ServiceTypeLoadBalancer)
+	defer func() {
+		jig.ChangeServiceType(svc.Namespace, svc.Name, v1.ServiceTypeClusterIP, framework.LoadBalancerCreateTimeoutDefault)
+		Expect(cs.CoreV1().Services(svc.Namespace).Delete(svc.Name, nil)).NotTo(HaveOccurred())
+	}()
+	ingressIP := framework.GetIngressPoint(&svc.Status.LoadBalancer.Ingress[0])
+	port := strconv.Itoa(int(svc.Spec.Ports[0].Port))
+	lbIpPort := net.JoinHostPort(ingressIP, port)
+
+	execPodName := framework.CreateExecPodOrFail(cs, ns, "execpod-affinity", nil)
+	defer func() {
+		framework.Logf("Cleaning up the exec pod")
+		err := cs.CoreV1().Pods(ns).Delete(execPodName, nil)
+		Expect(err).NotTo(HaveOccurred())
+	}()
+	execPod, err := cs.CoreV1().Pods(ns).Get(execPodName, metav1.GetOptions{})
```

Comment:
Ahhh, I just realized in kube-proxy iptables mode we redirect all pod -> exnternal VIP traffic to the Service's ClusterIP instead...So for LB service, testing affinity from a pod doesn't actually cover the case we wanted.
https://github.com/kubernetes/kubernetes/blob/f6eed81f211d929592ddb236535f306a277ff372/pkg/proxy/iptables/proxier.go#L1126-L1138

We might need to change the LB test to send request from the test VM instead (so it would be external traffic). Let's sync-up later :)

Reply:
Just committed changes. Test failed locally due to network latency. Here's part of the log.

```
Mar 15 12:12:23.287: INFO: Waiting up to 20m0s for service "service" to have a LoadBalancer
Mar 15 12:12:59.379: INFO: Testing HTTP reachability of http://35.225.221.27:80
Mar 15 12:13:04.379: INFO: Got error testing for reachability of http://35.225.221.27:80: Get http://35.225.221.27:80: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
Mar 15 12:13:06.380: INFO: Testing HTTP reachability of http://35.225.221.27:80
Mar 15 12:13:06.465: INFO: Got pod response: service-gzhzm
Mar 15 12:13:08.465: INFO: Testing HTTP reachability of http://35.225.221.27:80
Mar 15 12:13:08.549: INFO: Got pod response: service-xnrbm
Mar 15 12:13:08.662: INFO: Testing HTTP reachability of http://35.225.221.27:80
Mar 15 12:13:08.743: INFO: Got pod response: service-gzhzm
Mar 15 12:13:10.744: INFO: Testing HTTP reachability of http://35.225.221.27:80
Mar 15 12:13:11.021: INFO: Got pod response: service-xnrbm
Mar 15 12:13:12.744: INFO: Testing HTTP reachability of http://35.225.221.27:80
Mar 15 12:13:17.744: INFO: Got error testing for reachability of http://35.225.221.27:80: Get http://35.225.221.27:80: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
Mar 15 12:13:19.744: INFO: Testing HTTP reachability of http://35.225.221.27:80
Mar 15 12:13:20.857: INFO: Got pod response: service-gzhzm
Mar 15 12:13:22.744: INFO: Testing HTTP reachability of http://35.225.221.27:80
Mar 15 12:13:22.827: INFO: Got pod response: service-xnrbm
Mar 15 12:13:24.744: INFO: Testing HTTP reachability of http://35.225.221.27:80
Mar 15 12:13:24.826: INFO: Got pod response: service-gzhzm
Mar 15 12:13:26.744: INFO: Testing HTTP reachability of http://35.225.221.27:80
Mar 15 12:13:26.833: INFO: Got pod response: service-xnrbm
Mar 15 12:13:28.744: INFO: Testing HTTP reachability of http://35.225.221.27:80
Mar 15 12:13:28.826: INFO: Got pod response: service-gzhzm
Mar 15 12:13:30.744: INFO: Testing HTTP reachability of http://35.225.221.27:80
Mar 15 12:13:30.827: INFO: Got pod response: service-xnrbm
Mar 15 12:13:32.744: INFO: Testing HTTP reachability of http://35.225.221.27:80
Mar 15 12:13:37.744: INFO: Got error testing for reachability of http://35.225.221.27:80: Get http://35.225.221.27:80: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
Mar 15 12:13:39.746: INFO: Testing HTTP reachability of http://35.225.221.27:80
Mar 15 12:13:44.747: INFO: Got error testing for reachability of http://35.225.221.27:80: Get http://35.225.221.27:80: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
Mar 15 12:13:44.748: INFO: Testing HTTP reachability of http://35.225.221.27:80
Mar 15 12:13:49.748: INFO: Got error testing for reachability of http://35.225.221.27:80: Get http://35.225.221.27:80: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
Mar 15 12:13:49.748: INFO: Could not reach HTTP service through 35.225.221.27:80 after 2s: <nil>
<Test failed here>
```
It seems that it takes more than 2s from local to connect to the LB.
Please take a look and see if this is an issue. Thx.

Reply:
Figured out what is going on. Upon the session affinity update on service, corresponding target pool on GCP will be recreated, which introduces service downtime:
https://github.com/kubernetes/kubernetes/blob/bef0838a21a38ee341eaee79ae57d05e940d9d9a/pkg/cloudprovider/providers/gce/gce_loadbalancer_external.go#L697-L722

We are hitting the 60 seconds timeout waiting for the target pool to be recreated and take effect. Suggest we use `LoadBalancerPollTimeout` for the LB case.